xarray/core/common.py

```diff
@@ -938,6 +938,50 @@ def ones_like(other, dtype=None):
     return full_like(other, 1, dtype)
 
 
+def structured_like(data, template):
+    """Return a new object structured in the same way as the given object.
+
+    Parameters
+    ----------
+    data : array_like
+        Data to fill the new object. Must have same shape as template.
+    template : DataArray, or Variable
+        The reference object from which to get labels.
+
+    Returns
+    -------
+    out : same as template
+        New object with dimensions, attributes, coordinates and name
+        copied from template.
+
+    Examples
+    --------
+
+    >>> array = xr.DataArray([1, 2, 3], dims='x', coords={'x': ['a', 'b', 'c']})
+    >>> xr.structured_like([0.1, 0.2, 0.3], array)
+    <xarray.DataArray (x: 3)>
+    array([0.1, 0.2, 0.3])
+    Coordinates:
+    * x        (x) <U1 'a' 'b' 'c'
+    """
+    from .dataarray import DataArray
+    from .variable import Variable, as_compatible_data
+
+    if not (isinstance(template, DataArray) or isinstance(template, Variable)):
```

Comment:
Optional: `isinstance` can check against tuples of types all at the same time, so you could write: `if not isinstance(template, (Variable, DataArray)):

Reply:
nice!